File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.java
class Solution {
    private int[] findNLE(int[] arr, int n) {
        int[] nle = new int[n];
        Stack<Integer> st = new Stack<>();
        for(int i = n-1; i >= 0; i--) {
            while(!st.isEmpty() && arr[i] > arr[st.peek()]) st.pop();
            if(st.isEmpty()) nle[i] = n;
            else nle[i] = st.peek();
            st.push(i);
        }

        return nle;
    }
    private int[] findPLE(int[] arr, int n) {
        int[] ple = new int[n];
        Stack<Integer> st = new Stack<>();
        for(int i = 0; i < n; i++) {
            while(!st.isEmpty() && arr[i] >= arr[st.peek()]) st.pop();
            if(st.isEmpty()) ple[i] = -1;
            else ple[i] = st.peek();
            st.push(i);
        }

        return ple;
    }
    private long sumSubarrayMax(int[] arr) {
        int n = arr.length;
        int[] nle = findNLE(arr, n);
        int[] ple = findPLE(arr, n);

        long ans = 0;
        for(int i = 0; i < n; i++){
            int right = nle[i] - i;
            int left = i - ple[i];

            ans = ans + (long) right * left * arr[i];
        }

        return ans;
    }

    private int[] findNSE(int[] arr, int n){
        int[] nse = new int[n];
        Stack<Integer> st = new Stack<>();

        for(int i = n-1; i >= 0; i--){
            while(!st.empty() && arr[i] < arr[st.peek()]) st.pop();
            if(st.empty()) nse[i] = n;
            else nse[i] = st.peek();

            st.push(i);
        }

        return nse;
    }

    private int[] findPSE(int[] arr, int n){
        int[] pse = new int[n];
        Stack<Integer> st = new Stack<>();

        for(int i = 0; i < n; i++){
            while(!st.isEmpty() && arr[i] <= arr[st.peek()]) st.pop();
            if(st.empty()) pse[i] = -1;
            else pse[i] = st.peek();

            st.push(i);
        }

        return pse;
    }

    private long sumSubarrayMins(int[] arr) {
        int n = arr.length;
        int[] nse = findNSE(arr, n);
        int[] pse = findPSE(arr, n);

        long ans = 0;
        for(int i = 0; i < n; i++){
            int right = nse[i] - i;
            int left = i - pse[i];

            ans = (ans + (long) right * left * arr[i]);
        }

        return ans;
    }

    public long subArrayRanges(int[] nums) {
        return (long) sumSubarrayMax(nums) - sumSubarrayMins(nums);
    }
}

File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.cpp
class Solution {
private:
    vector<int> findNLE(vector<int>& arr, int n){
        stack<int> st;
        vector<int> nle(n, n);
        for(int i = n-1; i >= 0; i--){
            while(!st.empty() && arr[i] >= arr[st.top()]) st.pop();
            if(st.empty()) nle[i] = n;
            else nle[i] = st.top();
            st.push(i);
        }
        return nle;
    }
    vector<int> findPLE(vector<int>& arr, int n){
        stack<int> st;
        vector<int> ple(n, -1);
        for(int i = 0; i < n; i++){
            while(!st.empty() && arr[i] > arr[st.top()]) st.pop();
            if(st.empty()) ple[i] = -1;
            else ple[i] = st.top();
            st.push(i);
        }
        return ple;
    }
    long long sumSubarrayMaxs(vector<int>& arr) {
        int n = arr.size(), MOD = 1e9+7;
        vector<int> nle = findNLE(arr, n);
        vector<int> ple = findPLE(arr, n);
        long long ans = 0;
        for(int i = 0; i < n; i++){
            int right = (nle[i] - i);
            int left = (i - ple[i]);
            ans = ans + (right * left * 1LL * arr[i]);
        }

        return ans;
    }
    vector<int> findNSE(vector<int>& arr, int n){
        stack<int> st;
        vector<int> nse(n, -1);
        for(int i = n-1; i >= 0; i--){
            while(!st.empty() && arr[i] <= arr[st.top()]) st.pop();
            if(st.empty()) nse[i] = n;
            else nse[i] = st.top();
            st.push(i);
        }
        return nse;
    }
    vector<int> findPSE(vector<int>& arr, int n){
        stack<int> st;
        vector<int> pse(n, -1);
        for(int i = 0; i < n; i++){
            while(!st.empty() && arr[i] < arr[st.top()]) st.pop();
            if(st.empty()) pse[i] = -1;
            else pse[i] = st.top();
            st.push(i);
        }
        return pse;
    }
    long long sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        vector<int> nse = findNSE(arr, n);
        vector<int> pse = findPSE(arr, n);
        long long ans = 0;
        for(int i = 0; i < n; i++){
            int right = (nse[i] - i);
            int left = (i - pse[i]);
            ans = ans + (right * left * 1LL * arr[i]);
        }

        return ans;
    }
public:
    long long subArrayRanges(vector<int>& nums) {
        return sumSubarrayMaxs(nums) - sumSubarrayMins(nums);
    }
};